API: Adds transformer support in ColumnTransformer.remainder

thomasjpfan opened the pull request:
#### Reference Issues/PRs
Fixes #11210.

#### What does this implement/fix? Explain your changes.
Adds transformer support in the `ColumnTransformer.remainder` parameter.

thomasjpfan committed:
API: Adds transformer support in remainder

amueller commented:
awesome! flake8 is unhappy though.

amueller commented on sklearn/compose/_column_transformer.py:
```diff
@@ -97,6 +101,11 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
         Keys are transformer names and values are the fitted transformer
         objects.
 
+    remainder_transformers_ : list or None
```
why plural? that's only one, right? at least according to the docstring for ``remainder``.

thomasjpfan committed:
RFC: Uses a single remainder transformer

thomasjpfan committed:
TST: Improves coverage

thomasjpfan committed:
ENH: Support remainder transformer in make

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into issue-11210

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -227,14 +243,19 @@ def _validate_transformers(self):
-        if self.remainder not in ('drop', 'passthrough'):
+        is_transformer = ((hasattr(self.remainder, "fit")
+                           or hasattr(self.remainder, "fit_transform"))
+                          and hasattr(self.remainder, "transform"))
```
There is a small possibility that an estimator will not have `transform` until after fitting...

thomasjpfan replied:
I used the negation of condition in `_validate_transformers`:

```python
if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
        hasattr(t, "transform")):
    raise TypeError("All estimators should implement fit and "
                    "transform, or can be 'drop' or 'passthrough' "
                    "specifiers. '%s' (type %s) doesn't." %
                    (t, type(t)))
```

I am curious. Which estimator defines `transform` dynamically after calling `fit`?  🤔

thomasjpfan replied:
The documentation in `ColumnTransformer` forces transformers to have  `transform` defined:
```
        transformer : estimator or {'passthrough', 'drop'}
            Estimator must support `fit` and `transform`. Special-cased
            strings 'drop' and 'passthrough' are accepted as well, to
            indicate to drop the columns or to pass them through untransformed,
            respectively.
```
It should be okay to apply this restriction to `remainder` as well.

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -55,7 +55,8 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
             Like in Pipeline and FeatureUnion, this allows the transformer and
             its parameters to be set using ``set_params`` and searched in grid
-            search.
+            search. The name 'remainder' can not be used if the ``remainder``
```
I think we should outlaw the name remainder in all cases. In fact, I think `_BaseComposition._validate_names` enforces this.

jnothman reviewed (commented):
I wonder if the implementation here could be greatly simplified by appending remainder to `transformers_` and to `_iter`.

thomasjpfan commented:
As long as we can make 'remainder' an invalid transformer name, I think putting remainder into transformers_ would work.

thomasjpfan commented:
Now that I have tried to move `remainder` to transformer in `_iter`, I think it would be more complicated. There would have to be code to pull out the trained remainder estimator to update the `ColumnTransformer.remainder` object. 

The name `remainder` is already restricted in `_valid_names`, since `remainder` is a parameter in `ColumnTransformer`:
```python
invalid_names = set(names).intersection(self.get_params(deep=False))
if invalid_names:
    raise ValueError('Estimator names conflict with constructor '
                     'arguments: {0!r}'.format(sorted(invalid_names)))
```

This also forces me to name `remainder` to be `remainder_transformer` when it is placed in `transformers` in `_iter`.

thomasjpfan committed:
DOC: update

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into issue-11210

thomasjpfan committed:
RFC: Uses passed in deep parameter

thomasjpfan committed:
RFC: Uses iter for remainder training

thomasjpfan commented:
I updated this PR with an implementation where `_iter` is used to train `remainder`. Here are some details about the implementation:

1. It cleans up `fit_transform` and `transform`. `self._passthrough` does not need to be checked anymore. If `remainder` is 'passthrough' or an estimator, `_iter` will iterate through them. If `remainder` is 'drop', it will be ignored `_iter`.

2. `remainder_transformer` is used as the name of the remainder transformer in `_iter`.

3. `remainder_` is updated by looking for the name `remainder_transformer` (in `_update_fitted_transformers`).

4. `remainder_` is only added during `_iter`. It does not change `transformers_`. This keeps the external interface where: `len(transformers) == len(transformers_)`.

jnothman commented:
In master, 'remainder' is already not allowed as a transformer name:
```
>>> from sklearn.compose import ColumnTransformer
>>> from sklearn.preprocessing import OneHotEncoder
>>> ColumnTransformer([('remainder', OneHotEncoder(), [0])])
ColumnTransformer(n_jobs=1, remainder='passthrough', transformer_weights=None,
         transformers=[('remainder', OneHotEncoder(categorical_features='all', dtype=<class 'numpy.float64'>,
       handle_unknown='error', n_values='auto', sparse=True), [0])])
>>> ColumnTransformer([('remainder', OneHotEncoder(), [0])]).fit([[1]])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/joel/repos/scikit-learn/sklearn/compose/_column_transformer.py", line 361, in fit
    self._validate_transformers()
  File "/Users/joel/repos/scikit-learn/sklearn/compose/_column_transformer.py", line 215, in _validate_transformers
    self._validate_names(names)
  File "/Users/joel/repos/scikit-learn/sklearn/utils/metaestimators.py", line 68, in _validate_names
    'arguments: {0!r}'.format(sorted(invalid_names)))
ValueError: Estimator names conflict with constructor arguments: ['remainder']
```

jnothman commented:
> This keeps the external interface where: len(transformers) == len(transformers_).

I don't think there's any problem changing this as long as ColumnTransformer has not yet been released. If we can get this patch into the upcoming release, I'd much prefer the simplicity of appending the remainder transformer/passthrough to `transformers_`.

thomasjpfan committed:
RFC: Adds remainder to transformers_

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -55,7 +55,8 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
             Like in Pipeline and FeatureUnion, this allows the transformer and
             its parameters to be set using ``set_params`` and searched in grid
-            search.
+            search. The name 'remainder_transformer' can not be used if the
```
I don't think this is necessary. The name `remainder` can't be used in any case.

thomasjpfan committed:
RFC: Adds all remainder types to transformers_

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -167,7 +171,14 @@ def get_params(self, deep=True):
         params : mapping of string to any
             Parameter names mapped to their values.
         """
-        return self._get_params('_transformers', deep=deep)
```
You shouldn't need to change this. But it seems you've stumbled across a limitation in `_BaseComposition.get_params()`. It should be using `super(_BaseComposition, self).get_params(deep=deep)` instead of `super(_BaseComposition, self).get_params(deep=deep)` to handle the case that there are nested parameters not merely from the collection of estimators.

With that fixed, `set_params` works on `remainder` too.

jnothman replied:
You're welcome to submit a fix for that issue here, or in a separate PR.

thomasjpfan replied:
Update this PR with this fix.

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -227,23 +237,25 @@ def _validate_transformers(self):
-                self._passthrough = None
-        else:
-            self._passthrough = None
+        self._remainder = ('remainder_transformer', self.remainder,
```
just call it 'remainder'

thomasjpfan replied:
Does not pass `_validate_names` if this is named `remainder`. This is because `self.remainder` is defined in `__init__`.

jnothman replied:
So then this needs to be excluded when invoking `_validate_names`

thomasjpfan replied:
I would also prefer to name it `remainder`.

jnothman replied:
This is definitely not a use-case I'd foreseen for `_BaseComposition`, so if there's a more elegant way to structure `_BaseComposition`, feel free to change it!

thomasjpfan replied:
I think updating `_invalid_names` to let this through, would introduce an error when calling `set_params(remainder__{}={})` since there would be a `remainder` in transformer and a `self.remainder`.

thomasjpfan replied:
I will see what I can come up with for `_BaseComposition`.

jnothman replied:
> I will see what I can come up with for _BaseComposition.

Don't go out of your way. Only if it seems justified by your needs.

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -267,15 +279,16 @@ def get_feature_names(self):
         feature_names = []
         for name, trans, _, _ in self._iter(fitted=True):
-            if trans == 'drop':
+            if trans == 'drop' or name == 'remainder_transformer':
```
why should remainder be excluded from feature names?

thomasjpfan replied:
Good point. I updated the PR to remove the restriction.

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -294,7 +307,12 @@ def _update_fitted_transformers(self, transformers):
+        transformer_iter = chain(self.transformers, [self._remainder])
+
+        for name, old, column in transformer_iter:
+            if column is None:
```
why?

thomasjpfan replied:
If you are speaking of `transformer_iter`: It is used to iterate over the same number of items that `_iter. In `_iter`:

```python
transformers = chain(self.transformers, [self._remainder])
```

If you are speaking of `if column is None`: `_iter` also skips the transformer when column is None.

jnothman replied:
> If you are speaking of `if column is None`: `_iter` also skips the transformer when column is None.

Where do you see this logic?

thomasjpfan replied:
I updated `_iter` to include this:

```python
for name, trans, column in transformers:
    if column is None:
        continue
```

This condition happens when there are no `self._remainder=(_, _, None)`, which occurs when `passthrough==None` or when there are no remaining columns left.

jnothman replied:
well, maybe it's best to check that `name == 'remainder'` too. It would make the code clearer.

thomasjpfan committed:
TST: More checks on length

thomasjpfan commented:
If there are any remaining columns, the `remainder` will always be added to`transformers_`. This also includes `remainder='drop'` or `remainder='passthrough'`. Tests for the `len(transformers_)` were added to assert this behavior.

thomasjpfan committed:
RFC: Includes remainder in get_features

thomasjpfan committed:
BUG: Fixes #11332

thomasjpfan commented:
Mind as well fix #11332 as well. 😄

thomasjpfan committed:
RFC: Uses deep in get_params

jnothman commented:
> Mind as well fix #11332 as well

I wouldn't make a point of doing so here. It's definitely orthogonal to this change.

jnothman commented:
> RFC: Uses deep in get_params

This requires a test. Although testing `ColumnTransformer([...], remainder=SomeEst()).get_params()` would, IMO, be sufficient.

thomasjpfan committed:
TST: Add test for get_params

thomasjpfan committed:
STY: Space

thomasjpfan committed:
REV: Removes #11332 fix

thomasjpfan committed:
RFC: Adds name check

thomasjpfan commented:
The big issue remaining is to figure out how to avoid using `remainder_transformer` and use `remainder` as the remainder name in `transformers_`. 🤔

thomasjpfan committed:
RFC: Use remainder as name in transformers_

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -227,23 +232,24 @@ def _validate_transformers(self):
+        cols = []
+        for _, _, columns in self.transformers:
+            cols.extend(_get_column_indices(X, columns))
+        passthrough = sorted(list(set(range(n_columns)) - set(cols)))
```
let's give this a better name: `passthrough` -> `remaining_idx`

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -294,7 +301,12 @@ def _update_fitted_transformers(self, transformers):
+        transformer_iter = chain(self.transformers, [self._remainder])
+
+        for name, old, column in transformer_iter:
+            if column is None and name == 'remainder':
```
I think this should be done above at the `chain` stage.

jnothman reviewed (commented):
Rather than add a param to `_iter`, I'd personally prefer `_validate_names(zip(*self.transformers)[0])` and not using `_iter` to validate names. But I can see why the current version might also be construed as clearer.

jnothman reviewed (commented):
Please also update the documentation of the `transformers_` attribute.

thomasjpfan committed:
RFC: Checks remainder first

thomasjpfan committed:
DOC: Update transformers_

thomasjpfan commented:
PR updated with:

1. `_validate_transformers` checks `self.transformers` and does not call `_iter` anymore.
2. `self._remainder` is checked first before adding it to the chain.

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -90,7 +93,10 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
         The collection of fitted transformers as tuples of
-        (name, fitted_transformer, column).
+        (name, fitted_transformer, column). If there are any remaining
+        columns, a tuple of the form:
```
how about "the final element is a tuple of the form" and dropping "will be..."?

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -90,7 +93,10 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
     transformers_ : list
         The collection of fitted transformers as tuples of
-        (name, fitted_transformer, column).
+        (name, fitted_transformer, column). If there are any remaining
```
While you're here, you might note that "fitted_transformer" can be "drop".

jnothman commented on sklearn/compose/_column_transformer.py:
```diff
@@ -90,7 +93,10 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
     ----------
     transformers_ : list
```
perhaps specify the length here as len(transformers) or len(transformers) + 1

thomasjpfan committed:
DOC: Update

jnothman commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -571,3 +600,129 @@ def test_column_transformer_remainder_pandas(key):
+                           remainder=SparseMatrixTrans())
+
+    X_trans = ct.fit_transform(X_array)
+    assert_true(sparse.issparse(X_trans))
```
With the adoption of pytest, we are phasing out use of test helpers `assert_equal`, `assert_true`, etc. Please use bare `assert` statements, e.g. `assert x == y`, `assert not x`, etc.

You also don't need `assert_dict_equal`, just `assert d1 == d2`.

thomasjpfan replied:
I have a habit of keeping to the same style as the file I am working with. 

Since I am changing the new tests to bare asserts, do you mind if I go through the whole test file and change everything to bare asserts?

jnothman replied:
To avoid making it harder to review, don't do it here. I don't mind this coming in a separate PR, given that there isn't yet much concurrent work on this file. Don't replace the array comparison assertions. Mostly the assert_true/false/equal are worth removing

jnothman reviewed (commented):
Do you have a test case where you check the value of `transformers_[-1]`? Do you have a test case where there are no remaining columns?

Do you think that if there are no remaining columns but a remainder transformer is given, we should raise a warning?

thomasjpfan committed:
TST: Adds remainder checks

RFC: Use bare asserts

thomasjpfan committed:
TST: No remaining cols with remainder transformer

thomasjpfan commented:
For the case where there is no remaining columns and a remainder estimator, I do not think there needs to be a warning. A warning would have to direct a user to a preferred usage option. In our case it would either be to change `remainder` to `drop` or `passthrough`. I do not think those options are better than the initial estimator.

thomasjpfan committed:
REV: Go back to previous asserts

jnothman reviewed (commented):
Please also see if this can be incorporated into one of the examples in `examples/compose` or if it should be mentioned in the user guide.

thomasjpfan committed:
RFC: Does not need to check remainder

The remainder is already a part of `_iter`, thus it will be checked in the for loop.

thomasjpfan committed:
DOC: Adds remainder info to user guide

jorisvandenbossche commented on doc/modules/compose.rst:
```diff
@@ -448,6 +452,39 @@ as a list of multiple items, an integer array, a slice, or a boolean mask.
+
+The ``remainder`` parameter can be set to an estimator to transform the
+remaining rating columns. The transformed values are appended to the end of
+transformation::
```
"end of **the** transformation" 
(although I find 'transformation" a bit strange, but not directly a better suggestion)

thomasjpfan replied:
I used transformations because `ColumnTransformer` accepts `transformers` as a parameter. The "transformers" creates a "transformation", and the remainder is appended to it. 🤔

jorisvandenbossche commented on sklearn/compose/_column_transformer.py:
```diff
@@ -90,7 +93,13 @@ class ColumnTransformer(_BaseComposition, TransformerMixin):
         The collection of fitted transformers as tuples of
-        (name, fitted_transformer, column).
+        (name, fitted_transformer, column). `fitted_transformer` can be an
+        estimator or 'drop'. If there are remaining columns, the final
```
or 'passthrough' ?

jorisvandenbossche commented on sklearn/compose/_column_transformer.py:
```diff
@@ -227,23 +238,23 @@ def _validate_transformers(self):
 
     def _validate_remainder(self, X):
         """Generate list of passthrough columns for 'remainder' case."""
```
maybe update the docstring here (it was already not fully up to date, my fault)

jorisvandenbossche commented on sklearn/utils/metaestimators.py:
```diff
@@ -23,7 +23,7 @@ def __init__(self):
 
     def _get_params(self, attr, deep=True):
-        out = super(_BaseComposition, self).get_params(deep=False)
+        out = super(_BaseComposition, self).get_params(deep=deep)
```
This was needed to get the tests passing? (just to be sure this change is covered by the tests)

thomasjpfan replied:
This was updated to support `ColumnTransformer.get_params('remainder__{}')` where the  remainder is a estimator. This is tested with `test_column_transformer_get_set_params_with_remainder`.

jorisvandenbossche replied:
OK!

jorisvandenbossche commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -509,35 +542,47 @@ def test_column_transformer_remainder():
     assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both[:, ::-1])
+    assert len(ct.transformers_) == 2
+    assert ct.transformers_[-1][0] == 'remainder'
+    assert_array_equal(ct.transformers_[-1][2], [0])
```
do you test somewhere that in this case the transformer added to `transformers_` is the string 'passthrough' ? (and the same for drop)

jorisvandenbossche commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -571,3 +617,149 @@ def test_column_transformer_remainder_pandas(key):
+
+    assert_array_equal(ct.fit_transform(X_array), X_res_both)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both)
+    assert len(ct.transformers_) == 2
```
should we test here more explicitly the full content of `ct.trnasformers_[-1]` ? (you already test it partly in many other places, but here it seems appropriate to test the full content)

thomasjpfan replied:
I went through all the tests to fully test all the elements of `ct.transformers_[-1]`.

jorisvandenbossche replied:
Personally, I think that doing that everywhere is a bit overkill compared to a specific test to do it, but OK :-)

jorisvandenbossche commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -571,3 +617,149 @@ def test_column_transformer_remainder_pandas(key):
+    assert_raise_message(ValueError, "1D data passed to a transformer",
+                         col_trans.fit, X_array)
+    assert_raise_message(ValueError, "1D data passed to a transformer",
+                         col_trans.fit_transform, X_array)
```
I don't fully understand the reason for this test (it seems a bit superfluous because remainder is not related with the StandardScaler errorring on the 1D data)

jorisvandenbossche commented on sklearn/compose/tests/test_column_transformer.py:
```diff
@@ -571,3 +617,149 @@ def test_column_transformer_remainder_pandas(key):
+
+def test_column_transformer_no_estimators_get_params():
+    params = ColumnTransformer([], remainder=StandardScaler()).get_params()
+    assert params['remainder__with_mean']
```
Should this test also call `fit` to cover the change in `_validate_transformers` (to deal with empty list) ?

jorisvandenbossche reviewed (commented):
Thanks a lot for this PR! 

Really nice PR, only added a bunch of minor comments.

thomasjpfan committed:
DOC: Fix

thomasjpfan committed:
TST: Increases coverage

thomasjpfan committed:
STY: Whitespace

jorisvandenbossche reviewed (approved):
Thanks for the updates!

jnothman reviewed (approved):
Thanks, @thomasjpfan. Great work!
